index.js

```diff
@@ -220,28 +145,110 @@ module.exports.add = function add (file, options, cb) {
 				try {
 					sourceMap = JSON.parse(stripBom(data));
 				} catch (err) {}
-
-				if (sourceMap) {
-					return fixImportedSourceMap();
-				}
-				mapsLoaded();
+				callback();
 			};
 
 			fs.readFile(mapFile, 'utf8', function (err, data) {
 				if (err) {
 					if (options.debug) {
 						console.log(PLUGIN_NAME + '-add: Can\'t read map file :' + mapFile);
 					}
-					return mapsLoaded();
+					return callback();
 				}
 				sourceMapLoaded(data);
 			});
 
+		};
+
+		var asyncTasks = [fixImportedSourceMap];
+		if (sourceMap) {
+			sourceMap = sourceMap.toObject();
+			// sources in map are relative to the source file
+			sourcePath = path.dirname(file.path);
+			fileContent = convert.removeComments(fileContent);
+		} else {
+			asyncTasks.unshift(loadSourceMap);
 		}
+		async.waterfall(asyncTasks, callback);
 
-	} else {
-		mapsLoaded();
+	};
+
+	var mapsLoaded = function (callback) {
+
+		if (!sourceMap && options.identityMap) {
+			var fileType = path.extname(file.path);
+			var source = unixStylePath(file.relative);
+			var generator = new SourceMapGenerator({ file: source });
+			if (fileType === '.js') {
+				var tokenizer = acorn.tokenizer(fileContent, { locations: true });
+				while (true) {
+					var token = tokenizer.getToken();
+
+					if (token.type.label === 'eof') {
+						break;
+					}
+					var mapping = {
+						original: token.loc.start,
+						generated: token.loc.start,
+						source: source,
+					};
+					if (token.type.label === 'name') {
+						mapping.name = token.value;
+					}
+					generator.addMapping(mapping);
+				}
+				generator.setSourceContent(source, fileContent);
+				sourceMap = generator.toJSON();
+			} else if (fileType === '.css') {
+				var ast = css.parse(fileContent, { silent: true });
+				var registerTokens = function (ast) {
+					if (ast.position) {
+						generator.addMapping({
+							original: ast.position.start,
+							generated: ast.position.start,
+							source: source,
+						});
+					}
+					for (var key in ast) {
+						if (key !== 'position') {
+							if (Object.prototype.toString.call(ast[key]) === '[object Object]') {
+								registerTokens(ast[key]);
+							} else if (ast[key].constructor === Array) {
+								for (var i = 0; i < ast[key].length; i++) {
```

Comment:
Can you use functional iterators instead of forEach loops? Maps cleaner/less code.

This specific case would be one line: `ast[key].forEach(registerTokens)`